server/src/main/java/org/uiautomation/ios/wkrdp/WebKitNotificationListener.java

```diff
@@ -27,162 +35,278 @@
+    try {
+      applicationRegistrationLock.lock();
+      registeredApplications.add(applicationConnectedMessage.getApplication());
+      if (!isDriverSuccessfullyNotified()) {
+        log.log(Level.WARNING,
+            "No connectable applications found in ApplicationConnectedMessage, will wait before signalling driver");
+      }
+    } finally {
+      applicationRegistrationLock.unlock();
+    }
+  }
+
+  private boolean isDriverSuccessfullyNotified() {
+    for (WebkitApplication application : registeredApplications) {
+      if (application.isConnectableByWkrdProtocol()) {
+        driver.setApplications(registeredApplications);
         sync.signalSimSentApps();
+        return true;
       }
+    }
+    return false;
+  }
 
+  private boolean isValidForProcessing(ApplicationSentListingMessage applicationSentListingMessage) {
+    if (!applicationSentListingMessage.isPagesAvailable()) {
+      if (log.isLoggable(Level.INFO)) {
+        log.log(Level.INFO,
+            "ApplicationSentListingMessage '_rpc_applicationSentListing:' received with no pages, skipping processing");
+      }
+      return false;
     }
+    if (!isConnectableByWkrdProtocol(applicationSentListingMessage.getApplicationIdentifier())) {
+      if (log.isLoggable(Level.INFO)) {
+        log.log(Level.INFO, "Application id: " + applicationSentListingMessage.getApplicationIdentifier()
+            + " not allowed for processing");
+      }
+      return false;
+    }
+    return true;
+  }
 
-    if (message instanceof ApplicationSentListingMessage) {
-      ApplicationSentListingMessage m = (ApplicationSentListingMessage) message;
+  private boolean isConnectableByWkrdProtocol(String applicationIdentifier) {
+    int attempts = 2;
+    try {
+      return checkForAvailabilityIn(applicationIdentifier, attempts)
+          && getApplication(applicationIdentifier).isConnectableByWkrdProtocol();
+    } catch (InterruptedException e) {
+      if (log.isLoggable(Level.FINE)) {
+        log.log(Level.FINE, "InterruptedException while checking availability.");
+      }
+    } catch (IllegalStateException illegalStateException) {
+      log.log(Level.WARNING, "Unknown error, application with identifier: " + applicationIdentifier
+          + ", is not retrievable from list of registered applications");
+    }
+    return false;
+  }
 
-      List<WebkitPage> messagePages = m.getPages();
-      List<WebkitPage> driverPages = driver.getPages();
-      boolean equals = WebkitPage.equals(messagePages, driverPages);
+  private boolean checkForAvailabilityIn(String applicationIdentifier, int times) throws InterruptedException {
+    while (times > 0) {
+      for (WebkitApplication application : registeredApplications) {
+        if (application.getBundleId().equals(applicationIdentifier)) {
+          if (log.isLoggable(Level.FINE)) {
+            log.log(Level.FINE, "application available in registeredApplications");
```

Comment:
How about "registered applications" to be consistent with the message on line 180 ? :)